### PR TITLE
(zipkin) adds span serializer

### PR DESCRIPTION
**Pull request by jcchavezs**
Description: This PR adds the serializer interface to have better cohesion in the spans serializer
Risk Level: Low, since the default is still HTTP-JSON v1 based on https://github.com/apache/incubator-zipkin-api/blob/v0.2.1/zipkin-api.yaml.
Testing: Unit testing, manual integration test with real Zipkin collector server.
Docs Changes: none yet
Release Notes:

### CODE REVIEW

**Comment by jmarantz** [comment]
@jcchavezs needs master merge; that may fix the 'format' errors at least.

File: source/extensions/tracers/zipkin/span_buffer.h [hunk 1]
```diff
@@ -17,14 +20,19 @@ class SpanBuffer {
    * Constructor that creates an empty buffer. Space needs to be allocated by invoking
    * the method allocateBuffer(size).
    */
-  SpanBuffer() {}
+  SpanBuffer(const envoy::config::trace::v2::ZipkinConfig::CollectorEndpointVersion version)
```
Comment (htuch) [inline source/extensions/tracers/zipkin/span_buffer.h]: Should these constructor args be `const&`?

File: source/extensions/tracers/zipkin/span_buffer.h [hunk 2]
```diff
@@ -54,14 +62,13 @@ class SpanBuffer {
...
-   * each JSON in the array corresponds to one Zipkin span.
+   * @return the contents of the buffer as string
    */
-  std::string toStringifiedJsonArray();
+  std::string serialize(SpanSerializer& serializer)
 
-private:
-  // We use a pre-allocated vector to improve performance
-  std::vector<Span> span_buffer_;
+      private :
```
Comment (htuch) [inline source/extensions/tracers/zipkin/span_buffer.h]: Weird spacing..

File: source/extensions/tracers/zipkin/zipkin.pb.cc [hunk 3]
```diff
@@ -0,0 +1,2705 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (htuch) [inline source/extensions/tracers/zipkin/zipkin.pb.cc]: How come you are checking in generated code? Ideally we use our `*_proto_library` rules to generate this on-the-fly, we don't have any pregenerated stubs elsewhere in Envoy.

File: source/extensions/tracers/zipkin/zipkin_core_types.h [hunk 4]
```diff
@@ -183,7 +189,8 @@ class Annotation : public ZipkinBase {
    *
    * @return a stringified JSON.
    */
-  const std::string toJson() override;
+  const std::string toJson(const envoy::config::trace::v2::ZipkinConfig::CollectorEndpointVersion
```
Comment (htuch) [inline source/extensions/tracers/zipkin/zipkin_core_types.h]: All these args should be `const&`

File: test/extensions/tracers/zipkin/span_buffer_test.cc [hunk 5]
```diff
@@ -101,13 +103,13 @@ TEST(ZipkinSpanBufferTest, sizeConstructorEndtoEnd) {
...
-                               "}]";
+                               "}]";*/
   EXPECT_EQ(2ULL, buffer.pendingSpans());
-  EXPECT_EQ(expected_json_array_string, buffer.toStringifiedJsonArray());
+  /*EXPECT_EQ(expected_json_array_string, buffer.toStringifiedJsonArray());*/
 
   buffer.clear();
   EXPECT_EQ(0ULL, buffer.pendingSpans());
-  EXPECT_EQ("[]", buffer.toStringifiedJsonArray());
+  /*EXPECT_EQ("[]", buffer.toStringifiedJsonArray());*/
```
Comment (htuch) [inline test/extensions/tracers/zipkin/span_buffer_test.cc]: ?

File: test/extensions/tracers/zipkin/zipkin_core_types_test.cc [hunk 6]
```diff
@@ -484,7 +504,7 @@ TEST(ZipkinCoreTypesSpanTest, defaultConstructor) {
                 R"({"key":"http.return_code","value":"400",)"
                 R"("endpoint":{"ipv4":"192.168.1.2","port":3306,)"
                 R"("serviceName":"my_service_name"}}]})",
-            span.toJson());
+            span.toJson(envoy::config::trace::v2::ZipkinConfig::HTTP_JSON_V1));
```
Comment (htuch) [inline test/extensions/tracers/zipkin/zipkin_core_types_test.cc]: Are there any new tests?

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.cc [hunk 7]
```diff
@@ -122,25 +126,100 @@ Tracing::SpanPtr Driver::startSpan(const Tracing::Config& config, Http::HeaderMa
...
+ReporterImpl::NewInstance(Driver& driver, Event::Dispatcher& dispatcher,
+                          const std::string& collector_endpoint,
+                          const envoy::config::trace::v2::ZipkinConfig::CollectorEndpointVersion
+                              collector_endpoint_version) {
+  SpanSerializer span_serializer;
+
+  switch (collector_endpoint) {
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_JSON span_serializer =
+      new HTTPSpanSerializer() break;
+      case envoy::config::trace::v2::ZipkinConfig::HTTP_PROTO span_serializer =
```
Comment (htuch) [inline source/extensions/tracers/zipkin/zipkin_tracer_impl.cc]: All of this definitely needs `fix_format`..

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.cc [hunk 8]
```diff
@@ -122,25 +126,100 @@ Tracing::SpanPtr Driver::startSpan(const Tracing::Config& config, Http::HeaderMa
...
   span_buffer_.allocateBuffer(min_flush_spans);
 
   enableTimer();
 }
 
-ReporterPtr ReporterImpl::NewInstance(Driver& driver, Event::Dispatcher& dispatcher,
-                                      const std::string& collector_endpoint) {
-  return ReporterPtr(new ReporterImpl(driver, dispatcher, collector_endpoint));
+ReporterPtr
+ReporterImpl::NewInstance(Driver& driver, Event::Dispatcher& dispatcher,
```
Comment (htuch) [inline source/extensions/tracers/zipkin/zipkin_tracer_impl.cc]: `newInstance` etc.

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.cc [hunk 9]
```diff
@@ -122,25 +126,100 @@ Tracing::SpanPtr Driver::startSpan(const Tracing::Config& config, Http::HeaderMa
...
+                          const std::string& collector_endpoint,
+                          const envoy::config::trace::v2::ZipkinConfig::CollectorEndpointVersion
+                              collector_endpoint_version) {
+  SpanSerializer span_serializer;
+
+  switch (collector_endpoint) {
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_JSON span_serializer =
```
Comment (htuch) [inline source/extensions/tracers/zipkin/zipkin_tracer_impl.cc]: Does this compile?

File: source/extensions/tracers/zipkin/zipkin.proto [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+syntax = "proto3";
+
+package zipkin.proto3;
```
Comment (htuch) [inline source/extensions/tracers/zipkin/zipkin.proto]: Why isn't this sourced from an external package?

**Comment by dio** [comment]
@jcchavezs, sorry. Is there any reason for not iterating on top of https://github.com/envoyproxy/envoy/pull/6985? I think you can also add these changes on top of https://github.com/envoyproxy/envoy/pull/6985, which builds fine.

**Pull request closed without merging**